### DISCUSSION TITLE
Is there a callback for all responses, not just the main frame's response?

### DISCUSSION THREAD

**Discussion by postmodern** [rubycdp/ferrum#277] [Q&A]
First off, Ferrum is an _amazing_ project! I am trying to figure out how to setup a callback to receive all responses (main frame + any additional background AJAX/XHR responses). Is this possible? Digging through the source code there is a private method called `subscribe_response_received` which uses a `Network.responseReceived` event hook, which does return a `params` `Hash`, but I can't seem to successfully map that back to the `Network::Request` and `Network::Response` objects.

**Comment by route** [comment]
There's https://github.com/rubycdp/ferrum#network which is an array of all request/response pairs.

**Reply by postmodern** [reply]
but is there a callback similar to `on(:request)` but that triggers when a response is returned for a request?

**Reply by route** [reply]
There's intercept https://github.com/rubycdp/ferrum#interceptoptions only

**Reply by postmodern** [reply]
Appears that what I want is the [`Network.responseReceived` event](https://chromedevtools.github.io/devtools-protocol/tot/Network/#event-responseReceived).

**Reply by route** [reply]
You can subscribe to it manually. Use .subscribe
